resources/providers/awslib/multi_region.go

```diff
@@ -0,0 +1,135 @@
+	Fetch(fetcher func(T) ([]AwsResource, error)) ([]AwsResource, error)
+	GetMultiRegionsClientMap() map[string]T
+}
+
+type DescribeCloudRegions interface {
+	DescribeRegions(ctx context.Context, params *ec2.DescribeRegionsInput, optFns ...func(*ec2.Options)) (*ec2.DescribeRegionsOutput, error)
+}
+
+type MultiRegionWrapper[T any] struct {
+	clients map[string]T
+}
+
+// NewMultiRegionClients is a utility function that is used to create a map of client instances of a given type T for multiple regions.
+func (w *MultiRegionWrapper[T]) NewMultiRegionClients(client DescribeCloudRegions, cfg awssdk.Config, factory func(cfg awssdk.Config) T, log *logp.Logger) *MultiRegionWrapper[T] {
+	var clientsMap = make(map[string]T, 0)
+	for _, region := range getRegions(client, log) {
+		cfg.Region = region
+		clientsMap[region] = factory(cfg)
+	}
+
+	wrapper := &MultiRegionWrapper[T]{
+		clients: clientsMap,
+	}
+
+	return wrapper
+}
+
+// Fetch retrieves resources from multiple regions concurrently using the provided fetcher function.
+func (w *MultiRegionWrapper[T]) Fetch(fetcher func(T) ([]AwsResource, error)) ([]AwsResource, error) {
+	var err error
+	var wg sync.WaitGroup
+	var mux sync.Mutex
+	var crossRegionResources []AwsResource
+
+	if w.clients == nil {
+		return nil, errors.New("multi region clients have not been initialize")
+	}
+
+	for region, client := range w.clients {
+		wg.Add(1)
+		go func(client T, region string) {
+			defer wg.Done()
+			results, fetchErr := fetcher(client)
+			if fetchErr != nil {
+				err = fmt.Errorf("Fail to retrieve aws resources for region: %s, error: %v, ", region, fetchErr)
+			}
+
+			mux.Lock()
+			crossRegionResources = append(crossRegionResources, results...)
+			mux.Unlock()
+		}(client, region)
+	}
+
+	wg.Wait()
+	return crossRegionResources, err
+}
+
+func (w *MultiRegionWrapper[T]) GetMultiRegionsClientMap() map[string]T {
+	return w.clients
+}
+
+// GetRegions will initialize the singleton instance and perform the API request to retrieve the regions list only once, even if the function is called multiple times.
+// Subsequent calls to the function will return the stored regions list without making another API request.
+// In case of a failure the function returns the default region.
+func getRegions(client DescribeCloudRegions, log *logp.Logger) []string {
+	log.Debug("GetRegions starting...")
+	var mu sync.Mutex
+	var initErr error
+
+	mu.Lock()
+	defer mu.Unlock()
+	once.Do(func() {
+		log.Debug("Get aws regions for the first time")
+		awsRegions = &cachedRegions{}
+
+		output, err := client.DescribeRegions(context.Background(), nil)
+		if err != nil {
+			initErr = fmt.Errorf("failed DescribeRegions: %w", err)
+			once = &sync.Once{} // reset singleton upon error
+			return
+		}
+
+		for _, region := range output.Regions {
+			awsRegions.enabledRegions = append(awsRegions.enabledRegions, *region.RegionName)
+		}
+	})
```

Comment:
A race condition can happen here as well:
```
	if initErr != nil {
		log.Errorf("Error: %v, init client only for the default region: %s", initErr, DefaultRegion)
		awsRegions.enabledRegions = append(awsRegions.enabledRegions, DefaultRegion)
	}
```
Even though we're initializing a set with the same value so it will have no effect, wdyt?